File: scripts/refresh_index_and_sitemap.py
#!/usr/bin/env python3
"""
Deduplicate articles.json (by canonical URL / slug) and regenerate sitemap.xml.
Stdlib only — no API keys. Run from repo root:
  python scripts/refresh_index_and_sitemap.py
"""
from __future__ import annotations

import json
import os
import re
import sys
from datetime import datetime

REPO_ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
MAX_ARTICLES = 140

STOPWORDS = {
    "a", "an", "and", "are", "as", "at", "be", "by", "for", "from", "how", "in",
    "into", "is", "it", "of", "on", "or", "that", "the", "this", "to", "with",
    "guide", "tutorial", "complete", "explained", "basics", "beginner", "beginners",
    "cybersecurity", "security", "attack", "attacks", "vulnerability", "vulnerabilities",
    "learn", "using", "use", "what", "why", "works",
}

STRONG_TOKENS = {
    "nmap", "wireshark", "burp", "sqlmap", "metasploit", "hydra", "hashcat", "aircrack",
    "gobuster", "netcat", "nikto", "john", "password", "social", "engineering",
    "owasp", "ctf", "kali", "firewalls", "firewall", "tor", "proxychains",
}


def _article_key(a: dict) -> str:
    url = (a.get("url") or "").strip()
    if url:
        return url
    slug = (a.get("slug") or "").strip()
    cat = (a.get("category") or "").strip()
    if slug and cat:
        return f"/{cat}/{slug}.html"
    return slug


def dedupe_articles(articles: list) -> list[dict]:
    """One entry per URL/slug; keep the row with the latest date string."""
    best: dict[str, dict] = {}
    for a in articles:
        if not isinstance(a, dict):
            continue
        key = _article_key(a)
        if not key:
            continue
        d = a.get("date") or ""
        prev = best.get(key)
        if prev is None or d >= (prev.get("date") or ""):
            best[key] = a
    out = sorted(best.values(), key=lambda x: x.get("date", ""), reverse=True)
    return out


def _tokens(title: str) -> set[str]:
    raw = re.findall(r"[a-z0-9]+", (title or "").lower())
    return {t for t in raw if len(t) > 2 and t not in STOPWORDS}


def _strong_signature(title: str) -> set[str]:
    t = (title or "").lower()
    tokens = _tokens(t)
    sig = set(tokens & STRONG_TOKENS)
    if "sql" in t and "injection" in t:
        sig.add("__sqli__")
    if "xss" in t or "cross-site" in t or "cross site" in t:
        sig.add("__xss__")
    if "home" in t and "lab" in t:
        sig.add("__home_lab__")
    if "social" in t and "engineering" in t:
        sig.add("__social_engineering__")
    return sig


def _jaccard(a: set[str], b: set[str]) -> float:
    if not a or not b:
        return 0.0
    return len(a & b) / len(a | b)


def prune_similar_articles(articles: list[dict]) -> list[dict]:
    """
    Keep one canonical piece per repeated concept for non-CVE categories.
    Preference order: has description -> newest date.
    """
    ranked = sorted(
        articles,
        key=lambda a: (1 if (a.get("description") or "").strip() else 0, a.get("date", "")),
        reverse=True,
    )

    kept: list[dict] = []
    by_cat_tokens: dict[str, list[set[str]]] = {}
    by_cat_sig: dict[str, set[str]] = {}

    for a in ranked:
        cat = (a.get("category") or "").strip().lower()
        title = (a.get("title") or "").strip()
        if not title:
            continue

        # CVE items are naturally distinct by CVE ID; keep all.
        if cat == "cve":
            kept.append(a)
            continue

        tset = _tokens(title)
        sig = _strong_signature(title)
        seen_sig = by_cat_sig.setdefault(cat, set())
        seen_tokens = by_cat_tokens.setdefault(cat, [])

        if sig and (sig & seen_sig):
            continue
        if any(_jaccard(tset, prev) >= 0.58 for prev in seen_tokens):
            continue

        kept.append(a)
        seen_sig.update(sig)
        seen_tokens.append(tset)

    # Freshest first in resulting index
    kept.sort(key=lambda x: x.get("date", ""), reverse=True)
    return kept


def write_sitemap(articles: list[dict]) -> None:
    """Same structure as generate_article.update_sitemap (dedupe by loc)."""
    base = "https://howcanihack.com"
    lines = [
        '<?xml version="1.0" encoding="UTF-8"?>',
        '<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">',
        f"  <url><loc>{base}/</loc><changefreq>daily</changefreq><priority>1.0</priority></url>",
        f"  <url><loc>{base}/tutorials/</loc><changefreq>weekly</changefreq><priority>0.9</priority></url>",
        f"  <url><loc>{base}/news/</loc><changefreq>daily</changefreq><priority>0.9</priority></url>",
        f"  <url><loc>{base}/cve/</loc><changefreq>daily</changefreq><priority>0.8</priority></url>",
        f"  <url><loc>{base}/certifications/</loc><changefreq>weekly</changefreq><priority>0.8</priority></url>",
        f"  <url><loc>{base}/tools/</loc><changefreq>weekly</changefreq><priority>0.8</priority></url>",
        f"  <url><loc>{base}/beginner/</loc><changefreq>weekly</changefreq><priority>0.8</priority></url>",
        f"  <url><loc>{base}/about.html</loc><changefreq>monthly</changefreq><priority>0.5</priority></url>",
        f"  <url><loc>{base}/contact.html</loc><changefreq>monthly</changefreq><priority>0.5</priority></url>",
        f"  <url><loc>{base}/editorial-policy.html</loc><changefreq>monthly</changefreq><priority>0.5</priority></url>",
        f"  <url><loc>{base}/write-for-us.html</loc><changefreq>monthly</changefreq><priority>0.5</priority></url>",
        f"  <url><loc>{base}/privacy.html</loc><changefreq>monthly</changefreq><priority>0.4</priority></url>",
        f"  <url><loc>{base}/disclaimer.html</loc><changefreq>monthly</changefreq><priority>0.4</priority></url>",
    ]

    url_lastmod: dict[str, str] = {}
    for a in articles:
        url = base + (a.get("url") or "")
        if not url or url == base:
            continue
        lastmod = a.get("date") or datetime.now().strftime("%Y-%m-%d")
        prev = url_lastmod.get(url)
        if prev is None or lastmod > prev:
            url_lastmod[url] = lastmod

    for url in sorted(url_lastmod.keys()):
        lastmod = url_lastmod[url]
        lines.append(
            f'  <url><loc>{url}</loc><lastmod>{lastmod}</lastmod>'
            f'<changefreq>weekly</changefreq><priority>0.7</priority></url>'
        )

    lines.append("</urlset>")
    out_path = os.path.join(REPO_ROOT, "sitemap.xml")
    with open(out_path, "w", encoding="utf-8") as f:
        f.write("\n".join(lines) + "\n")


def main() -> int:
    os.chdir(REPO_ROOT)
    path = os.path.join(REPO_ROOT, "articles.json")
    if not os.path.isfile(path):
        print("No articles.json — writing sitemap with section pages only.")
        write_sitemap([])
        print("🗺️  sitemap.xml updated")
        return 0

    with open(path, encoding="utf-8") as f:
        try:
            raw = json.load(f)
        except json.JSONDecodeError as e:
            print(f"Invalid JSON: {e}")
            return 1

    if not isinstance(raw, list):
        print("articles.json must be a list")
        return 1

    before = len(raw)
    deduped = dedupe_articles(raw)
    curated = prune_similar_articles(deduped)
    curated = curated[:MAX_ARTICLES]

    with open(path, "w", encoding="utf-8") as f:
        json.dump(curated, f, indent=2)
        f.write("\n")

    write_sitemap(curated)
    print(
        f"📋 articles.json: {before} → {len(deduped)} rows (exact dedupe)"
        f" → {len(curated)} rows (semantic curation)"
    )
    print("🗺️  sitemap.xml regenerated")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/refresh_index_and_sitemap.mjs
import { readFileSync, writeFileSync } from "fs";
import { dirname, join } from "path";
import { fileURLToPath } from "url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const REPO_ROOT = join(__dirname, "..");
const MAX_ARTICLES = 140;

const STOPWORDS = new Set([
  "a", "an", "and", "are", "as", "at", "be", "by", "for", "from", "how", "in",
  "into", "is", "it", "of", "on", "or", "that", "the", "this", "to", "with",
  "guide", "tutorial", "complete", "explained", "basics", "beginner", "beginners",
  "cybersecurity", "security", "attack", "attacks", "vulnerability", "vulnerabilities",
  "learn", "using", "use", "what", "why", "works"
]);

const STRONG_TOKENS = new Set([
  "nmap", "wireshark", "burp", "sqlmap", "metasploit", "hydra", "hashcat", "aircrack",
  "gobuster", "netcat", "nikto", "john", "password", "social", "engineering",
  "owasp", "ctf", "kali", "firewalls", "firewall", "tor", "proxychains"
]);

const articleKey = (a) => {
  const url = (a.url || "").trim();
  if (url) return url;
  const slug = (a.slug || "").trim();
  const cat = (a.category || "").trim();
  if (slug && cat) return `/${cat}/${slug}.html`;
  return slug;
};

const tokens = (title) => {
  const raw = (title || "").toLowerCase().match(/[a-z0-9]+/g) || [];
  return new Set(raw.filter((t) => t.length > 2 && !STOPWORDS.has(t)));
};

const strongSignature = (title) => {
  const t = (title || "").toLowerCase();
  const tok = tokens(t);
  const sig = new Set([...tok].filter((x) => STRONG_TOKENS.has(x)));
  if (t.includes("sql") && t.includes("injection")) sig.add("__sqli__");
  if (t.includes("xss") || t.includes("cross-site") || t.includes("cross site")) sig.add("__xss__");
  if (t.includes("home") && t.includes("lab")) sig.add("__home_lab__");
  if (t.includes("social") && t.includes("engineering")) sig.add("__social_engineering__");
  return sig;
};

const jaccard = (a, b) => {
  if (!a.size || !b.size) return 0;
  const inter = [...a].filter((x) => b.has(x)).length;
  const union = new Set([...a, ...b]).size;
  return union ? inter / union : 0;
};

const dedupeExact = (rows) => {
  const best = new Map();
  for (const a of rows) {
    if (!a || typeof a !== "object") continue;
    const key = articleKey(a);
    if (!key) continue;
    const prev = best.get(key);
    if (!prev || (a.date || "") >= (prev.date || "")) best.set(key, a);
  }
  return [...best.values()].sort((x, y) => (y.date || "").localeCompare(x.date || ""));
};

const curateSemantic = (rows) => {
  const ranked = [...rows].sort((a, b) => {
    const ad = (a.description || "").trim() ? 1 : 0;
    const bd = (b.description || "").trim() ? 1 : 0;
    if (bd !== ad) return bd - ad;
    return (b.date || "").localeCompare(a.date || "");
  });

  const kept = [];
  const byCatTokens = new Map();
  const byCatSig = new Map();

  for (const a of ranked) {
    const cat = (a.category || "").trim().toLowerCase();
    const title = (a.title || "").trim();
    if (!title) continue;
    if (cat === "cve") {
      kept.push(a);
      continue;
    }

    const tset = tokens(title);
    const sig = strongSignature(title);
    const seenSig = byCatSig.get(cat) || new Set();
    const seenTokens = byCatTokens.get(cat) || [];

    if (sig.size && [...sig].some((x) => seenSig.has(x))) continue;
    if (seenTokens.some((p) => jaccard(tset, p) >= 0.58)) continue;

    kept.push(a);
    for (const s of sig) seenSig.add(s);
    seenTokens.push(tset);
    byCatSig.set(cat, seenSig);
    byCatTokens.set(cat, seenTokens);
  }

  return kept.sort((x, y) => (y.date || "").localeCompare(x.date || "")).slice(0, MAX_ARTICLES);
};

const writeSitemap = (articles) => {
  const base = "https://howcanihack.com";
  const lines = [
    '<?xml version="1.0" encoding="UTF-8"?>',
    '<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">',
    `  <url><loc>${base}/</loc><changefreq>daily</changefreq><priority>1.0</priority></url>`,
    `  <url><loc>${base}/tutorials/</loc><changefreq>weekly</changefreq><priority>0.9</priority></url>`,
    `  <url><loc>${base}/news/</loc><changefreq>daily</changefreq><priority>0.9</priority></url>`,
    `  <url><loc>${base}/cve/</loc><changefreq>daily</changefreq><priority>0.8</priority></url>`,
    `  <url><loc>${base}/certifications/</loc><changefreq>weekly</changefreq><priority>0.8</priority></url>`,
    `  <url><loc>${base}/tools/</loc><changefreq>weekly</changefreq><priority>0.8</priority></url>`,
    `  <url><loc>${base}/beginner/</loc><changefreq>weekly</changefreq><priority>0.8</priority></url>`,
    `  <url><loc>${base}/about.html</loc><changefreq>monthly</changefreq><priority>0.5</priority></url>`,
    `  <url><loc>${base}/contact.html</loc><changefreq>monthly</changefreq><priority>0.5</priority></url>`,
    `  <url><loc>${base}/editorial-policy.html</loc><changefreq>monthly</changefreq><priority>0.5</priority></url>`,
    `  <url><loc>${base}/write-for-us.html</loc><changefreq>monthly</changefreq><priority>0.5</priority></url>`,
    `  <url><loc>${base}/privacy.html</loc><changefreq>monthly</changefreq><priority>0.4</priority></url>`,
    `  <url><loc>${base}/disclaimer.html</loc><changefreq>monthly</changefreq><priority>0.4</priority></url>`
  ];

  const urlLastmod = {};
  const today = new Date().toISOString().slice(0, 10);
  for (const a of articles) {
    const url = base + (a.url || "");
    if (!url || url === base) continue;
    const lastmod = a.date || today;
    if (!urlLastmod[url] || lastmod > urlLastmod[url]) urlLastmod[url] = lastmod;
  }
  for (const url of Object.keys(urlLastmod).sort()) {
    lines.push(
      `  <url><loc>${url}</loc><lastmod>${urlLastmod[url]}</lastmod><changefreq>weekly</changefreq><priority>0.7</priority></url>`
    );
  }
  lines.push("</urlset>");
  writeFileSync(join(REPO_ROOT, "sitemap.xml"), lines.join("\n") + "\n", "utf8");
};

const path = join(REPO_ROOT, "articles.json");
const raw = JSON.parse(readFileSync(path, "utf8"));
const exact = dedupeExact(raw);
const curated = curateSemantic(exact);
writeFileSync(path, JSON.stringify(curated, null, 2) + "\n", "utf8");
writeSitemap(curated);
console.log(`articles.json: ${raw.length} -> ${exact.length} (exact) -> ${curated.length} (semantic)`);
console.log("sitemap.xml regenerated");
